### ISSUE TITLE
Space to points

### ISSUE THREAD

**Issue by WIPocket** [processing/p5.js#1817]
While using the "textToPoints" function, a space in the string returns a TypeError: "pathArray is undefined".

**Comment by limzykenneth** [comment]
@WIPocket `textToPoints` doesn't seemed to be a p5.js function, can you provide more details? If it is a library function, please raise the issue with the library maintainer, they can better help with your problem.

**Comment by WIPocket** [comment]
`TypeError: pathArray is undefined

pathToAbsolute()
 p5.js:27294
path2curve()
 p5.js:27380
pointAtLength()
 p5.js:27257
pathToPoints()
 p5.js:27113
[72]</p5.Font.prototype.textToPoints()
 p5.js:26852
getPosData()
 p5.text-dot.js:44
setup/usePattern()
 p5.text-dot.js:79
setup/<()`

Also yes, this is a P5.js function (from "p5.js", in "p5.min.js" it says "a" is undefined)
I got this copy of the library from Daniel Shiffmans github, but even when I used the cdn on cloudflare I got the same error (still only when a spave was in the string)
Thanks

**Comment by limzykenneth** [comment]
Right, I see. The `textToPoint` function is not showing up in the reference but it does exist in the source with some documentation, I'm not sure if it is meant to be an unfinished undocumented function or it simply is a mistake. Someone more familiar with p5.Font implementation will have to take this up for you, I can't be of much help at the moment I'm afraid.

**Comment by dhowe** [comment]
Yes, there was some discussion about whether this should be a documented or undocumented function. I think in the end we decided that it should be documented (to enable sketches like [this](https://rednoise.org/glyphflock/)). @lmccart, perhaps you can weigh in on this? Meantime, a fix for this issue is here: https://github.com/processing/p5.js/pull/1823

**Comment by WIPocket** [comment]
I looked at the commit and tryed it, but it still doesnt work:
No errors show up, but all points after the space are on x 0 (Rendering on the very left of my canvas)
Perhaps make the offset bigger and skip it? (using `continue;` in the for loop after "Making the space" by adding to the offset to be used from than on)

**Comment by dhowe** [comment]
I don't see this (check below). Can you include a simple test?

![image](https://cloud.githubusercontent.com/assets/737638/23332778/473c153c-fbba-11e6-851a-2c5e4a6ac265.png)

**Comment by lmccart** [comment]
I think it's worth documenting if that doesn't add any complication or issues.. are there any?

**Comment by dhowe** [comment]
do you mean documenting the function itself? it is documented inline, not sure what it takes to make it show up in the ref:

```
/**
 * Computes an array of points following the path for specified text
 *
 * @param  {String} txt     a line of text
 * @param  {Number} x        x-position
 * @param  {Number} y        y-position
 * @param  {Number} fontSize font size to use (optional)
 * @param  {Object} options  an (optional) object that can contain:
 *
 * <br>sampleFactor - the ratio of path-length to number of samples
 * (default=.25); higher values yield more points and are therefore
 * more precise
 *
 * <br>simplifyThreshold - if set to a non-zero value, collinear points will be
 * be removed from the polygon; the value represents the threshold angle to use
 * when determining whether two edges are collinear
 *
 * @return {Array}  an array of points, each with x, y, alpha (the path angle)
 */
```

**Comment by lmccart** [comment]
oh it just needs
```
@method textToPoints
...
at the top. I'll add it next chance I get if nobody else gets there first.

**Comment by dhowe** [comment]
@WIPocket any further info on the issue you mention above?

> all points after the space are on x 0

**Comment by WIPocket** [comment]
Most of the code is by Daniel Shiffman:
https://github.com/CodingTrain/Rainbow-Code/tree/master/challenges/CC_59_Steering_Text_Paths

All I did is added a function to "redraw" it.
(and I also changed the p5 function to the new code)

**Comment by dhowe** [comment]
Problem seems to be in your additions, as the basic code works ok:

![image](https://cloud.githubusercontent.com/assets/737638/23397444/80298ae4-fdd3-11e6-8c00-60af87b3d938.png)

Post it if you want feedback...

**Comment by WIPocket** [comment]
![tmp_12428-screenshot_2017-03-01-14-23-12-1033226420](https://cloud.githubusercontent.com/assets/18397681/23461587/d4cb08cc-fe8a-11e6-86c4-c7c523a51dd2.png)

This is the original sketch from Daniel Shiffman (sketch.js and vehicle.js)

**Comment by WIPocket** [comment]
![tmp_12428-screenshot_2017-03-01-14-23-12-1033226420](https://cloud.githubusercontent.com/assets/18397681/23461587/d4cb08cc-fe8a-11e6-86c4-c7c523a51dd2.png)

This is the original sketch from Daniel Shiffman (sketch.js and vehicle.js)

**Comment by dhowe** [comment]
Please include the actual code

**Comment by WIPocket** [comment]
`// Literally just a copy of Dainels scripts
var font;
var vehicles = [];

function preload() {
  font = loadFont('AvenirNextLTPro-Demi.otf');
  console.log("preload");
}

function setup() {
  createCanvas(600, 300);
  background(51);
  console.log("setup");
  // textFont(font);
  // textSize(192);
  // fill(255);
  // noStroke();
  // text('train', 100, 200);

  var points = font.textToPoints('tra in', 100, 200, 192, {
    sampleFactor: 0.25
  });

  for (var i = 0; i < points.length; i++) {
    var pt = points[i];
    var vehicle = new Vehicle(pt.x, pt.y);
    vehicles.push(vehicle);
    // stroke(255);
    // strokeWeight(8);
    // point(pt.x, pt.y);
  }
}
console.log("hw");
function draw() {
	//console.log("draw");
  background(51);
  for (var i = 0; i < vehicles.length; i++) {
    var v = vehicles[i];
    v.behaviors();
    v.update();
    v.show();
  }
}

function Vehicle(x, y) {
  this.pos = createVector(random(width), random(height));
  this.target = createVector(x, y);
  this.vel = p5.Vector.random2D();
  this.acc = createVector();
  this.r = 8;
  this.maxspeed = 10;
  this.maxforce = 1;
}

Vehicle.prototype.behaviors = function() {
  var arrive = this.arrive(this.target);
  var mouse = createVector(mouseX, mouseY);
  var flee = this.flee(mouse);

  arrive.mult(1);
  flee.mult(5);

  this.applyForce(arrive);
  this.applyForce(flee);
}

Vehicle.prototype.applyForce = function(f) {
  this.acc.add(f);
}

Vehicle.prototype.update = function() {
  this.pos.add(this.vel);
  this.vel.add(this.acc);
  this.acc.mult(0);
}

Vehicle.prototype.show = function() {
  stroke(255);
  strokeWeight(this.r);
  point(this.pos.x, this.pos.y);
}


Vehicle.prototype.arrive = function(target) {
  var desired = p5.Vector.sub(target, this.pos);
  var d = desired.mag();
  var speed = this.maxspeed;
  if (d < 100) {
    speed = map(d, 0, 100, 0, this.maxspeed);
  }
  desired.setMag(speed);
  var steer = p5.Vector.sub(desired, this.vel);
  steer.limit(this.maxforce);
  return steer;
}

Vehicle.prototype.flee = function(target) {
  var desired = p5.Vector.sub(target, this.pos);
  var d = desired.mag();
  if (d < 50) {
    desired.setMag(this.maxspeed);
    desired.mult(-1);
    var steer = p5.Vector.sub(desired, this.vel);
    steer.limit(this.maxforce);
    return steer;
  } else {
    return createVector(0, 0);
  }
}`

**Comment by shiffman** [comment]
I'm happy to help debug this. @WIPocket does the issue happen with this example:

https://github.com/CodingTrain/Rainbow-Code/tree/master/challenges/CC_59_Steering_Text_Paths

If so I can make a simpler version that only looks at the path points. It's quite likely that my example does nothing to account for spaces and maybe it needs to?

**Comment by dhowe** [comment]
I can't recreate the issue. Here is a mod of the sketch from @shiffman which allows text entry:

http://rednoise.org/sketches/textpaths/

![image](https://cloud.githubusercontent.com/assets/737638/23576741/ebe719e6-00e8-11e7-82e6-22de368b62c7.png)

**Comment by WIPocket** [comment]
I dont seem to have this issue on your site. Can it have anything to do with the fact I dont have it on a server (all local)?
Also I don't enter the option for more dots.
I will try to rewrite the code by linking to your version of the full p5.js code.

**Comment by GeoffLouw** [comment]
This seems to be font specific, some fonts works prefect.
I tried the same string with different fonts and only getting this error on some fonts.

**Comment by dhowe** [comment]
If you are using a character that is not in the font, this could happen. Can you post an example of the problem ?

**Comment by WIPocket** [comment]
I used `console.log("space");` in the if statenent when the glyph name is `space` and it showed in the dev console (it was a space). Also when I used a char thats not in the font (`》`) it showed a questionmark in a box.
So all my code does is it loads a font in preload and than loops in all the vectors. I made it write them to the page and when I used a space everything after it was `x == 0`.
Hope this helped a little, but I will try to run it with different fonts.
